### PR TITLE
display specific error messages for password change (issue #5549)

### PR DESCRIPTION
**Pull request by MothOnMars**
- added cuke tests for email preferences change

### CODE REVIEW

**Comment by MothOnMars** [comment]
This is my first PR for diaspora, so...be gentle!  Some notes on the changes:
* I changed the redirect following the update to a render to allow Rails to wrap the faulty fields in the ```fields_with_errors``` div and to pass the error messages to ```f.error_messages```. Most of the other changes were required to support that change.
* After the ```render``` allowed ```f.error_messages``` to be displayed, I realized that they were duplicated for each form on that page. I removed the duplicate error message lines for the sections that don't allow error-prone user input (such as the languages drop-down). 
* I was wondering why there are multiple forms for user changes on the same page. It might not be a bad idea to refactor the form and update code (which I'm happy to do, but I wanted to keep my first PR relatively lightweight).
* I am not generally a front-end coder, so please let me know if the CSS changes should be made elsewhere.
* The updated form is below. I used the standard bootstrap styling for the error fields, but the error messages could probably use some styling:
image

**Comment by jaywink** [comment]
Hi, thanks for your work, looks good apart from some comments.

I like the way error messages are shown, but I'd IMHO remove the "*Invalid Fields*" - it's kinda unnecessary as the next line explains the same. Also, maybe make the *"Correct the ...."* and the error lines red colour (or some shade of red we use elsewhere. That should be enough with the bootstrappy way of highlighting errors in the box itself too (wrapping it in red).

The other thing is that many pages seem to be broken, the header overlaps the top of the page for example and stream box has weird bootstrappy looking effects. Maybe because you included `bootstrap` which probably is not the way to go?

**Comment by Flaburgan** [comment]
Oh, it's nice to see you working on that but... that's not what I meant in #5549 :p 
I was talking about "reset password" when the user forgets it and fill his email to be able to set a new password, not changing password in the user settings.

To deal with the real reset password problem, I would wait for #5391 to be merged, I change a lot of things there.

**Comment by jaywink** [comment]
This PR can be merged though even though it's not related to #5549, when problems are fixed.

**Comment by MothOnMars** [comment]
Thanks @jaywink  & @Flaburgan for the review & comments. Sorry about the confusion about reset password vs. change password! I agree that it's probably better to address the reset password issue after #5391 is merged, but yeah, this can probably be merged after I change a couple of things.

@jaywink, I'm happy to remove the "Invalid Fields" line. I thought that was probably unnecessary, but I didn't want to make changes that would affect other pages. Colorizing "Correct the following errors..." would also affect those pages. However, it looks like the ```error_messages``` helper that includes that text is only used on a couple of other views:
```
$ gg "\.error_messages" app/views/ | uniq
app/views/aspects/new.haml:    = aspect.error_messages
app/views/publisher/_publisher_blueprint.html.haml:      = status.error_messages
app/views/publisher/_publisher_bootstrap.html.haml:      = status.error_messages
app/views/users/edit.html.haml:            = f.error_messages
app/views/users/edit.html.haml:                    = f.error_messages
app/views/users/edit.mobile.haml:    = f.error_messages
app/views/users/edit.mobile.haml:            = f.error_messages
app/views/users/privacy_settings.html.haml:        = f.error_messages
```
As long as no one minds that those changes will affect those other pages, I can remove that line.

Regarding the pages being broken, I'll try to sort out what when wrong.

**Comment by jhass** [comment]
Hey, #5391 was merged :)

**Comment by MothOnMars** [comment]
Ah, excellent. I haven't forgotten about this, I just suddenly got very busy. I'll make the remaining changes for this and check out the changes in #5391 .

**Comment by MothOnMars** [comment]
I've updated this PR with a few changes:
* removed the "Invalid Fields" header from the error messages (this will apply to all forms using the `error_messages` helper)
* colorized the error messages
* added cuke test and added error messages for password reset form
* removed styling for fields with errors. I ran into a number of issues when trying to extend Bootstrap's classes for field errors, and I decided to leave that to someone with more CSS fu, as it wasn't actually a requirement for this bug fix. We may want to add that as a UI enhancement after Bootstrap is upgraded to v3.

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -4,17 +4,17 @@
 
 class UsersController < ApplicationController
   before_action :authenticate_user!, :except => [:new, :create, :public, :user_photo]
-  layout ->(c) { request.format == :mobile ? "application" : "with_header_with_footer" }, only: [:privacy_settings, :edit]
+  before_action -> { @css_framework = :bootstrap }, only: [:privacy_settings, :edit, :update]
```
Comment (jhass) [inline app/controllers/users_controller.rb]: Rebase issue, the `@css_framework` hack is no longer needed.

File: app/views/users/edit.html.haml [hunk 2]
```diff
@@ -43,8 +43,9 @@
         .span12
           %h3 
             = t('.change_password')
-          = form_for 'user', :url => user_path, :html => { :method => :put } do |f|
-            = f.error_messages
+          = form_for @user, :url => user_path, :html => { :method => :put } do |f|
+            - if @user.errors.keys & [:password, :password_confirmation, :current_password]
+              = f.error_messages
```
Comment (jhass) [inline app/views/users/edit.html.haml]: `[:a] & [:b]` is `[]` which is truthy.

Comment (MothOnMars) [inline app/views/users/edit.html.haml]: Good call - thanks for catching that

File: Changelog.md [hunk 3]
```diff
@@ -146,7 +146,7 @@ diaspora.yml file**. The existing settings from 0.4.x and before will not work a
 * Fix overflow for headings, blockquotes and other elements [#5731](https://github.com/diaspora/diaspora/pull/5731)
 * Correct photo count on profile page [#5751](https://github.com/diaspora/diaspora/pull/5751)
 * Fix mobile sign up from an invitation [#5754](https://github.com/diaspora/diaspora/pull/5754)
-* Set max-width for tag following button on tag page [#5752](https://github.com/diaspora/diaspora/pull/5752)
+* Display error messages for failed password change [#5549](https://github.com/diaspora/diaspora/issues/5549)
```
Comment (svbergerem) [inline Changelog.md]: You accidentally removed one line from the changelog.

File: app/assets/stylesheets/new_styles/_forms.scss [hunk 4]
```diff
@@ -103,3 +103,7 @@ form.block-form {
     ::placeholder { text-transform: uppercase; }
   }
 }
+
+.form_errors {
+  color : #b94a48;
```
Comment (svbergerem) [inline app/assets/stylesheets/new_styles/_forms.scss]: Please use a color variable here. You could try `$red`. If that is not ok for you please add this color as a new variable in `colors.scss`.

Comment (MothOnMars) [inline app/assets/stylesheets/new_styles/_forms.scss]: will do

File: Changelog.md (hunk 3 above)
Comment (MothOnMars) [inline Changelog.md]: Whoops - rebase fail. Fixing...

File: app/assets/stylesheets/new_styles/_forms.scss (hunk 4 above)
Comment (svbergerem) [inline app/assets/stylesheets/new_styles/_forms.scss]: Another solution could be the `text-error` class [Bootstrap](http://getbootstrap.com/2.3.2/base-css.html#typography) has. Just put the error messages in an element with that class and the text color will be red.

**Comment by svbergerem** [comment]
Thank you!

**Pull request merged**